### PR TITLE
DBZ-188: Allow a debezium mysql connector to filter production of DML…

### PR DESCRIPTION
**Pull request by dasl-**
… events into kafka by the mysql UUID of the event

With GTIDs enabled, each transaction in the binlog contains a GTID event, which gives us access to the GTID of the transaction. The GTID has the following format: source_id:transaction_id, where source_id is the UUID of the mysql server the transaction was written to.

I propose to allow a debezium instance to be configured with a UUID pattern to check against before producing DML events into Kafka. Debezium would produce a DML event into kafka if and only if the UUID in the event's GTID matches the pattern with which debezium was configured.

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 1]
```diff
@@ -486,6 +486,18 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
...
+     * A comma-separated list of regular expressions that match source UUIDs of transactions for which we will produce
+     * DML events into Kafka. When DML events having UUIDs that do not match any of these regular expressions are
+     * processed, they will be ignored and not produced into Kafka.
+     */
+    public static final Field GTID_DML_SOURCE_INCLUDES = Field.create("gtid.dml_source.includes")
+                                                          .withDisplayName("Include GTID sources for DML production into Kafka")
+                                                          .withType(Type.LIST)
+                                                          .withWidth(Width.LONG)
+                                                          .withImportance(Importance.HIGH)
+                                                          .withDescription("We will only produce DML events into Kafka for transactions that were written on mysql servers with UUIDs included in this list.");
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Does it make sense to ignore DML events based upon the GTID source UUIDs defined in the `gtid.source.includes` or `gtid.source.excludes` properties?

**Comment by rhauch** [comment]
@dasl- thanks for the PR and suggested logic. Before we get into the nitty gritty of the changes, I'd like to first discuss the configuration. Could we simplify the configuration by having this enhancement make use of the existing GTID source filter? 

@criccomini any thoughts, since you're already using the existing GTID source filters?

**Comment by dasl-** [comment]
I'd be open to making use of the existing GTID source filter. I believe that would work fine for us, and it is simpler.

**Comment by rhauch** [comment]
@dasl- okay, so the question is whether the connector should *always* filter out events based upon the GTID source filter. If so, then no configuration property changes are required; if not, then your new configuration property probably should be a boolean and named something like `gtid.source.filter.events` and default to `true`??

@criccomini wdyt?

**Comment by dasl-** [comment]
makes sense to me! I'd be amenable to either approach.

**Comment by rhauch** [comment]
Until we hear from @criccomini, let's plan on having a configuration property that defaults to `true`.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 2]
```diff
@@ -101,6 +105,9 @@ public BinlogReader(String name, MySqlTaskContext context) {
         client.registerLifecycleListener(new ReaderThreadLifecycleListener());
         if (logger.isDebugEnabled()) client.registerEventListener(this::logEvent);
 
+        String gtidDmlSourceIncludes = context.config().getString(MySqlConnectorConfig.GTID_DML_SOURCE_INCLUDES);
+        gtidDmlSourceFilter = gtidDmlSourceIncludes != null ? Predicates.includes(gtidDmlSourceIncludes) : null;
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: The `MySqlTaskContext` has a `gtidSourceFilter()` method that returns a `Predicate<String>` that you could use. If we want a config option that specifies whether the events should be filtered with the GTID sources, then this could check that and if true just use the result of `gtidSourceFilter()`.

Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: 👍

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 3]
```diff
@@ -490,6 +504,9 @@ protected void handleInsert(Event event) throws InterruptedException {
             logger.debug("Skipping previously processed row event: {}", event);
             return;
         }
+        if (ignoreDmlEventByGtidSource) {
+            return;
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Probably should have a DEBUG log statement (similar to line 504) saying why it was skipped.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 4]
```diff
@@ -533,6 +550,9 @@ protected void handleUpdate(Event event) throws InterruptedException {
...
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Similar log message here; see [previous comment](https://github.com/debezium/debezium/pull/182/files#r100388528)

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 3 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: 👍 . Btw, how do I enable the debug logging? I am pretty new to java -- most of my development is in PHP.

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: See https://gitter.im/debezium/dev?at=58949a5de836bf7010c8566d, but in this case use `io.debezium.connector.mysql` for the context (which are names of Java packages or classes). I'll add a document to the website that goes into more specifics.

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Unless you're using the Docker images, then you just set the `LOG_LEVEL` environment variable to `DEBUG` (or `TRACE` or any other valid level).

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 5]
```diff
@@ -50,7 +52,7 @@
 
 /**
  * A component that reads the binlog of a MySQL server, and records any schema changes in {@link MySqlSchema}.
- * 
+ *
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Can you remove unrelated changes like this from your commits? Ideally, the pull request only changes the minimum number of lines. Let me know if you need an easy way to do this.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 6]
```diff
@@ -408,12 +415,19 @@ protected void handleGtidEvent(Event event) {
         String gtid = gtidEvent.getGtid();
         gtidSet.add(gtid);
         source.startGtid(gtid, gtidSet.toString()); // rather than use the client's GTID set
+        ignoreDmlEventByGtidSource = false;
+        if (gtidDmlSourceFilter != null && gtid != null) {
+            String uuid = gtid.substring(0, gtid.indexOf(":"));
+            if (!gtidDmlSourceFilter.test(uuid)) {
+                ignoreDmlEventByGtidSource = true;
+            }
+        }
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Shouldn't we set `ignoreDmlEventByGtidSource = false` also when the transaction is committed, around [this line](https://github.com/debezium/debezium/pull/182/files#diff-63435e78f18ae89b82241e3bf53289edL440)?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 7]
```diff
@@ -817,6 +836,17 @@ private static int validateGtidSetExcludes(Configuration config, Field field, Va
         return 0;
     }
 
+    private static int validateGtidSourceFilterDmlEvents(Configuration config, Field field, ValidationOutput problems) {
+        String includes = config.getString(GTID_SOURCE_INCLUDES);
+        String excludes = config.getString(GTID_SOURCE_EXCLUDES);
+        Boolean filterDmlEvents = config.getBoolean(GTID_SOURCE_FILTER_DML_EVENTS);
+        if (includes == null && excludes == null) {
+            problems.accept(GTID_SOURCE_FILTER_DML_EVENTS, filterDmlEvents , "Either gtid.source.includes or gtid.source.excludes must be specified.");
+            return 1;
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Should this problem be reported only if the boolean value of `filterDmlEvents` is true?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 5 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: will do, i have my editor (intellij) set to remove all trailing whitespace upon saving. will adjust.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java (hunk 7 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: good point, will fix.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 6 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: I'm not sure. I assumed every transaction has a GTID event so the value of `ignoreDmlEventByGtidSource` can be appropriately reset there. 

Perhaps the case you are pointing out is possible if there is a GTID transaction followed by a non-GTID transaction, but I'm not sure that's possible... in any case, maybe it's better safe than sorry. I will add that.

**Comment by criccomini** [comment]
Haven't fully read through code yet, but w.r.t. the config question, I agree that using the existing GTID filter config is better. I'm curious, what would be a use case for setting gtid.source.filter.events to false?

**Comment by rhauch** [comment]
@criccomini I think its equivalent to what you're doing now. But once this is in, I'm not sure whether the default for `gtid.source.filter.events` should remain `false`. I guess the question for you is whether your use case would work if `gtid.source.filter.events=true`?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 8]
```diff
@@ -408,12 +414,19 @@ protected void handleGtidEvent(Event event) {
         String gtid = gtidEvent.getGtid();
         gtidSet.add(gtid);
         source.startGtid(gtid, gtidSet.toString()); // rather than use the client's GTID set
+        ignoreDmlEventByGtidSource = false;
+        if (gtidDmlSourceFilter != null && gtid != null) {
+            String uuid = gtid.substring(0, gtid.indexOf(":"));
```
Comment (criccomini) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Is this going to work if GTIDs have more than one UUID in them? Not quite sure how DML GTIDs look--do they only ever have one UUID? It would make sense if so, but just want to confirm.

**Comment by criccomini** [comment]
@rhauch I believe it'll be fine for our use case. We just filter out the UUIDS for our replica servers, so we only maintain UUIDs for the upstream micro-service servers. Each DBZ instance only pays attention to the micro-service DB it's interested in, so I think this should work fine.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 8 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: AFAIK they only ever have one UUID because every transaction is only written to one mysql server.

Comment (criccomini) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: That's what I figured, just wanted to confirm since I've only ever seen the sever-level GTIDs which have multiple UUIDs.

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Do the `gitd` strings need to be trimmed? Again, just depends on whether they are already properly formatted. I believe that the `GtidSet` class is trimming internally, just to make sure inadvertent whitespace doesn't mess things up.

**Comment by rhauch** [comment]
@criccomini okay, in that case I think the default for `gtid.source.filter.events` should probably be true, since that likely makes more sense. @dasl-, if you agree, then the validation logic probably needs to be changed or removed altogether.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 8 above)
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: will trim.

**Comment by dasl-** [comment]
will update the default to true.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 9]
```diff
@@ -408,6 +414,13 @@ protected void handleGtidEvent(Event event) {
         String gtid = gtidEvent.getGtid();
         gtidSet.add(gtid);
         source.startGtid(gtid, gtidSet.toString()); // rather than use the client's GTID set
+        ignoreDmlEventByGtidSource = false;
+        if (gtidDmlSourceFilter != null && gtid != null) {
+            String uuid = gtid.trim().substring(0, gtid.indexOf(":"));
+            if (!gtidDmlSourceFilter.test(uuid)) {
```
Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: This seems like a fairly "hot" codepath. Is it ok performance wise to do a regex match everytime we have a gtid event?

**Comment by criccomini** [comment]
> I think the default for gtid.source.filter.events should probably be true

👍

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 9 above)
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: It would probably be better to use a `Set<String>` so that this is faster. Up until now, the filter was not used that frequently. When we create the filter in the `MySqlTaskContext` constructor, we could actually see if the include and exclude config values contain comma-separated UUIDs (rather than regex patterns). I can make that change if you prefer, or I can describe how I might go about doing it so you can give it a try.

Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Good idea. If the config values are non-regex, we can use a `Set`, otherwise fall back to the `Predicate` filter?

How will we determine if the config values are non-regex? Just see if all the characters in the config are alphanumeric or commas?

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: I'd just try parsing them using `UUID.fromString(str)`; any exceptions mean they are not literal UUIDs. I can't think of any regex expressions that might also be valid UUIDs, unless they are just UUID literals. Can you?

Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Nice, that should work. According to this a UUID can only consist of alphnumerics and hyphens: https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html#toString()

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java [hunk 10]
```diff
@@ -47,8 +48,38 @@ public MySqlTaskContext(Configuration config) {
         // Set up the GTID filter ...
         String gtidSetIncludes = config.getString(MySqlConnectorConfig.GTID_SOURCE_INCLUDES);
         String gtidSetExcludes = config.getString(MySqlConnectorConfig.GTID_SOURCE_EXCLUDES);
-        this.gtidSourceFilter = gtidSetIncludes != null ? Predicates.includes(gtidSetIncludes)
-                : (gtidSetExcludes != null ? Predicates.excludes(gtidSetExcludes) : null);
+        Boolean filterIncludesRegex = false;
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java]: I've submitted a different PR (#183) that improves the `Predicates` class to add support for UUID-based predicates. The benefit of my approach is that it handles any combination of UUID literal and regular expressions in the GTID source filter configuration properties, but it will return the most efficient predicate implementation (e.g., using a single set for all UUID literals, and only using that one set if no regular expressions are involved). And, it includes some unit tests.

If you concur, I can merge #183 and then you can rebase your branch on `master` and simplify your PR to only deal with using the filter and defining the new config property.

Comment (dasl-) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java]: 👍

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java]: @dasl- #183 is not merged.

**Comment by rhauch** [comment]
See #183 for an improved and more efficient handling of the GTID source filters.

**Comment by rhauch** [comment]
Superseded by #184
